internal/controller/apiextensions/composite/composition_render.go

```diff
@@ -80,26 +80,19 @@ func RenderFromJSON(o resource.Object, data []byte) error {
 	return nil
 }
 
-// RenderFromCompositePatches renders the supplied composed resource by applying
-// all patches that are _from_ the supplied composite resource.
-func RenderFromCompositePatches(cd resource.Composed, xr resource.Composite, p []v1.Patch) error {
+// RenderFromCompositeAndEnvironmentPatches renders the supplied composed
+// resource by applying all patches that are _from_ the supplied composite
+// resource or are from or to the supplied environment.
+func RenderFromCompositeAndEnvironmentPatches(cd resource.Composed, xr resource.Composite, e *Environment, p []v1.Patch) error {
 	for i := range p {
```

Comment:
nit: wdyt of rewriting it as follows?
```golang
func RenderFromCompositeAndEnvironmentPatches(cd resource.Composed, xr resource.Composite, e *Environment, p []v1.Patch) error {
	for i := range p {
		var src runtime.Object
		switch {
		case filterPatch(p[i], patchTypesFromXR()...):
			src = xr
		case e != nil && filterPatch(p[i], patchTypesFromToEnvironment()...):
			src = e
		default:
			continue
		}
		if err := ApplyToObjects(p[i], src, cd); err != nil {
			return errors.Wrapf(err, errFmtPatch, p[i].Type, i)
		}
	}
	return nil
}
```

Reply:
I think that is way to complicated and I would stick to `if`.

Reply:
With the original implementation it looks like both the `Apply` and `ApplyToObjects` will be called every time unnecessarily (which is actually the case), moving the filtering outside of the Apply calls makes it much more explicit in my opinion, but as I said it's just a nit, so not a blocker.